include/irr/asset/CTexturePacker.h

```diff
@@ -53,20 +55,119 @@ class CSquareAddressAllocator
+
+        uint32_t bufOffset = 0u;
+        for (uint32_t i = 0u; i < m_pageTable->getCreationParameters().mipLevels; ++i)
+        {
+            const uint32_t tilesPerLodDim = _tilesPerDim>>i;
+            const uint32_t regionSz = tilesPerLodDim*tilesPerLodDim*_numLayers*4u;
+            auto& region = (*regions)[i];
+            region.bufferOffset = bufOffset;
+            region.bufferImageHeight = 0u;
+            region.bufferRowLength = tilesPerLodDim*4u;
+            region.imageExtent = {tilesPerLodDim,tilesPerLodDim,1u};
+            region.imageOffset = {0,0,0};
+            region.imageSubresource.baseArrayLayer = 0u;
+            region.imageSubresource.layerCount = _numLayers;
+            region.imageSubresource.mipLevel = i;
+            region.imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+
+            bufOffset += regionSz;
+        }
+        auto buf = core::make_smart_refctd_ptr<ICPUBuffer>(bufOffset);
+
+        m_pageTable->setBufferAndRegions(std::move(buf), regions);
+    }
+
+    core::smart_refctd_dynamic_array<STexOffset> alloc(const ICPUImage* _img, const ICPUImage::SSubresourceRange& _subres)
+    {
+        auto extent = _img->getCreationParameters().extent;
+        uint32_t levelCount = 1u;
+        for (uint32_t i = 0u; (extent.width>>(_subres.baseMipLevel+i) > m_pgSzxy) || (extent.height>>(_subres.baseMipLevel+i) > m_pgSzxy); ++i)
+            levelCount = i;
+        levelCount = std::min(_subres.levelCount, levelCount);
+
+        auto offsets = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<STexOffset>>(levelCount);
+        for (uint32_t i = 0u; i < levelCount; ++i)
+        {
+            const uint32_t w = std::max(extent.width>>(_subres.baseMipLevel+i),1u);
+            const uint32_t h = std::max(extent.height>>(_subres.baseMipLevel+i),1u);
+            const uint32_t pgSqrSz = core::roundUpToPoT((std::max(w,h)+m_pgSzxy-1u)/m_pgSzxy);
+
+            const uint32_t addr = m_addrAlctr.alloc_addr(pgSqrSz);
+            (*offsets)[i] = (addr==CSquareAddressAllocator<uint32_t>::invalid_address) ? 
+                STexOffset::invalid_addr :
+                STexOffset(m_addrAlctr.unpackAddress_x(addr), m_addrAlctr.unpackAddress_y(addr), m_addrAlctr.unpackAddress_layer(addr));
+
+            //i didnt know where to put page table page table and filling code so it is here for now
+            uint32_t* const pgTab = reinterpret_cast<uint32_t*>(
+                reinterpret_cast<uint8_t*>(m_pageTable->getBuffer()->getPointer()) + m_pageTable->getRegions().begin()[i].bufferOffset
+                );
+            const uint32_t pitch = m_pageTable->getRegions().begin()[i].bufferRowLength;
+            const uint32_t offset = (pitch*(m_tilesPerDim>>i))*(*offsets)[i].layer() + (*offsets)[i].y()*pitch + (*offsets)[i].x();
```

Comment:
can't you get the offset and other stuff from your page table image's region spec?